### PR TITLE
Remove styled placeholder when input has value

### LINKED ISSUE #4867: TextInput renders styled placeholders even when there is a value
A TextInput that has a styled placeholder (`placeholder={<someJSX>}` rather than `placeholder="somestring"`) renders the placeholder even when the textinput has a value, at least until another render happens. 

### Expected Behavior

If a TextInput has a value (either from a value prop or by typing text in the input), the placeholder should not be shown.

### Actual Behavior

One first render the placeholder will be drawn whether or not the input has a value.
On next render, if the input has a value the placeholder will not be rendered. Note that typing in the text field does not cause a re-render so the typed text will overlap the placeholder.

### URL, screen shot, or Codepen exhibiting the issue

image

### Steps to Reproduce
```jsx
<Grommet>
    <h1>some content</h1>
    <TextInput
      data-testid="styled-placeholder"
      id="styled-placeholder"
      name="styled-placeholder"
      placeholder={<Text>placeholder text</Text>}
    />
</Grommet>
```

This happens because on first render, the inputRef is not defined yet and so this check in TextInput's render returns true even though a value was specified:
```javascript
const showStyledPlaceholder =
      placeholder &&
      typeof placeholder !== 'string' &&
      !(inputRef.current && inputRef.current.value);
```

There is also nothing that causes a re-render on typing text. Some onChange/setValue type of code is necessary to cause the input to re-render on input.

Note that there is a test case in TextInput-test that tries to make sure the placeholder doesn't get rendered but it only works if something causes the input to get rendered twice.
 
### Your Environment

- Grommet version: 2.16.2
- Browser Name and version: Chrome
- Operating System and version (desktop or mobile): Windows10

### PR DESCRIPTION
**Pull request by taysea**
#### What does this PR do?
Manages presence of StyledPlaceholder via state (showStyledPlaceholder). When `onChange` is called within the TextInput, it checks if the input will have a value or not. If yes, it removes the styled placeholder. If no, it keeps it.

This specifically addresses the case when a TextInput is not in a Form. When the TextInput is not in a Form, no rerender was happening when a value changed and therefore the styledplaceholder was not disappearing when the textinput had a value. A detailed explanation of this issue is [well documented here](https://github.com/grommet/grommet/issues/4867).

When a TextInput is inside a form, all worked correctly and continues to work correctly.

#### Where should the reviewer start?
https://github.com/grommet/grommet/issues/4867

#### What testing has been done on this PR?
Test in storybook with https://storybook.grommet.io/?path=/story/input-textinput--styled-placeholder by removing the `Form` tags. As you type, the placeholder should disappear and if there value is cleared, the placeholder should reappear.

#### How should this be manually tested?
Same as above.

#### What are the relevant issues?
Closes #4867 

#### Do the grommet docs need to be updated?
No.

#### Should this PR be mentioned in the release notes?
Yes. Removes custom JSX placeholder from TextInput when input has a value.

#### Is this change backwards compatible or is it a breaking change?
Backwards compatible.

### CODE REVIEW

File: src/js/components/TextInput/TextInput.js [hunk 1]
```diff
@@ -448,6 +450,11 @@ const TextInput = forwardRef(
               readOnly
                 ? undefined
                 : event => {
+                    // when TextInput is not contained in a Form, no re-render
+                    // will come from this onChange and remove the placeholder,
+                    // so we need to update state to ensure the styled
+                    // placholder only appears when there is no value
+                    setShowStyledPlaceholder(!event.target.value);
```
Comment (MikeKingdom) [inline src/js/components/TextInput/TextInput.js:457]: When I investigated this, it seems the onChange doesn't get called while the user is typing until they tab out or change focus. As a result the placeholder remains and the typed text overlaps it. Are you seeing something similar?

Comment (taysea) [reply src/js/components/TextInput/TextInput.js:457]: onChange was getting called as soon as they were typing. However, it wasn't causing a rerender for whatever reason (which made the placeholder remain). I think it has to do with the fact that `value` is looking to formContext and so if the TextInput isn't in a formContext it's not registering the value change as a state change. 

by converting showStyledPlaceholder to something that is controlled by TextInput state, it allows us to toggle when it's shown by checking if the event.target has a value or not. Let me know if this makes sense.

Comment (taysea) [reply src/js/components/TextInput/TextInput.js:457]: I was running this locally in Chrome. Let me confirm that it works across browsers.

Comment (taysea) [reply src/js/components/TextInput/TextInput.js:457]: It is working as expected on Chrome, Firefox, and Safari.

Comment (MikeKingdom) [reply src/js/components/TextInput/TextInput.js:457]: Looks great to me. I gave it a try and it definitely removes the placeholder on typing.

**Review by ericsoderberghp — APPROVED** [review]

**Pull request merged**